### PR TITLE
Add POST request functionality; relocate backend files; refactor frontend

### PR DESCRIPTION
**Pull request by smburdick**
* Backend: support POST request in Flask, move server files from library to their own directory
* Frontend: serialize plaquettes and qubits into JSON, add to request payload, component class refactoring

### CODE REVIEW

File: tqecjs/src/workspace/index.js
Comment (rryoung98) [inline tqecjs/src/workspace/index.js:97]: What's the difference between the .isTile in the class already?

**Review by rryoung98 — CHANGES_REQUESTED** [review]
Although structurally im a little hesitant to put the post request in the index.js, it's a good start

File: tqecjs/src/workspace/index.js
Comment (nelimee) [reply tqecjs/src/workspace/index.js:156]: The plaquette qubits should be positioned with respect to the plaquette local coordinate system:
image
If you provide something different, the backend will have to pre-process the coordinates. This is not an issue, we simply have to agree on the convention here.

File: tqecjs/src/workspace/index.js
Comment (smburdick) [reply tqecjs/src/workspace/index.js:97]: The reason why I added it here is that when I go to download, I traverse the workspace containers, and want to filter on those that are tiles. 
```js
		const tiles = workspace.children.filter((child) => child.isTile);
```

File: tqecjs/src/workspace/index.js
Comment (smburdick) [reply tqecjs/src/workspace/index.js:156]: You're right. Let's agree on the convention that the leftmost, uppermost qubit is at point (0,0) and other qubits in the plaquette are positioned relative to that.

Since the plaquette can be any convex hull, we do allow for negative x/y coordinates.

I'll have to think about how I want to implement this, but I think sorting and relabeling in function invoking the `POST` request is a good approach.

**Comment by smburdick** [comment]
> Although structurally im a little hesitant to put the post request in the index.js,

Can you elaborate on where it should be?

File: tqecjs/src/workspace/index.js
Comment (nelimee) [reply tqecjs/src/workspace/index.js]: Following the discussions in #34 I think the local coordinate system will eventually change. Let's keep it like that for the moment, but the origin will likely become the central qubit (labeled 5 in the picture above).

**Comment by nelimee** [comment]
Additionally to my comments, I have issues with #40 that introduced non-library files into the `tqec` library folder. Another PR might be the best place to discuss this, I do not know.
As a side note, it would be nice if I was a required reviewer for anything backend-related. I didn't have the time to comment on #40 before it was merged, and I had a few comments.

File: tqec/main.py
Comment (smburdick) [reply tqec/main.py:29]: Yes

File: tqec/main.py
Comment (smburdick) [reply tqec/main.py:1]: The server files are structured to accommodate Flask conventions. I'll move the server stuff to a different subdirectory of the root level of the repo (`tqec-backend`), and treat `tqec/` as a library, the latter of which I intended to do anyway.

File: tqec/main.py
Comment (smburdick) [reply tqec/main.py:37]: +1 to that. Have you considered a library like Pandas?

File: tqec/static/index.html
Comment (smburdick) [reply tqec/static/index.html:1]: Agreed, will move those out of `tqec/tqec/` altogether per my above comment.

**Comment by smburdick** [comment]
> Additionally to my comments, I have issues with #40 that introduced non-library files into the `tqec` library folder. Another PR might be the best place to discuss this, I do not know. As a side note, it would be nice if I was a required reviewer for anything backend-related. I didn't have the time to comment on #40 before it was merged, and I had a few comments.

In any future PR that modifies the backend, I will request your review.

File: tqec/main.py
Comment (nelimee) [reply tqec/main.py]: Not really, what does pandas have to do with that?

Comment (smburdick) [reply tqec/main.py]: I was suggesting using a (de-)serialization library for that, but I'll leave it up to you.

Comment (nelimee) [reply tqec/main.py]: I might try to use [`pydantic`](https://docs.pydantic.dev/latest/), because it seems to be the right tool and I want to learn about it, but a simple use of the standard `json` package and a few (~10) methods should be enough and will likely be:
1. easier and quicker to write,
2. easier to understand for newcomers, 
3. enough for our current purpose.

I am open to discussion on that part though, I never really made a JSON export/import interface to be used by external languages / other developers than myself, so I am a beginner on that subject.

**Comment by smburdick** [comment]
> Although structurally im a little hesitant to put the post request in the index.js, it's a good start and the buttons etc can be moved alter.

I believe I have addressed this concern by refactoring the buttons.

File: tqec/main.py
Comment (vtomole) [reply tqec/main.py:1]: The way i usually do this is put the server in a directory called `web` that's outside the lib directory.

**Comment by vtomole** [comment]
> I have issues with https://github.com/QCHackers/tqec/pull/40 that introduced non-library files into the tqec library folder.

We can easily move non-library file like this outside the `tqec` directory.

>  it would be nice if I was a required reviewer for anything backend-related

Ideally, the server setup should be independent of the back-end stuff you implement and thus shouldn't strictly need your review. Let's setup a process that makes this possible.

File: tqec/main.py
Comment (nelimee) [reply tqec/main.py:1]: That would be perfect for me, and also decouple the python library from the server backend implementation.

**Comment by smburdick** [comment]
> Ideally, the server setup should be independent of the back-end stuff you implement and thus shouldn't strictly need your review. Let's setup a process that makes this possible.

Perhaps we could each be assigned to one or more teams (eg, frontend, web, lib)? Since there are only a few of us working on this codebase, I have just been requesting @vtomole , @rryoung98 , @nelimee on PR’s, but I’m open to adding more structure, especially as more people join. I know we have teams already for the tasks, but I don’t know if the status of that has changed since our last discussion in the Group.

**Comment by nelimee** [comment]
> > Ideally, the server setup should be independent of the back-end stuff you implement and thus shouldn't strictly need your review. Let's setup a process that makes this possible.
> 
> Perhaps we could each be assigned to one or more teams (eg, frontend, web, lib)? Since there are only a few of us working on this codebase, I have just been requesting @vtomole , @rryoung98 , @nelimee on PR’s, but I’m open to adding more structure, especially as more people join. I know we have teams already for the tasks, but I don’t know if the status of that has changed since our last discussion in the Group.

I guess that we can still wait and see who is willing to help with the code before assigning teams. And maybe speak about the current state of the code next Wednesday (I can do that for the library part).

Also, I think that the backend server files should be moved in this PR (per its description). Once this is done, I am happy to validate the changes :)

**Comment by vtomole** [comment]
@smburdick My suggestion to add a `web` directory was more for repo organization not team organization.

**Comment by smburdick** [comment]
> @smburdick My suggestion to add a `web` directory was more for repo organization not team organization.

Understood. I agree with the directory name changes -- I'll rename `tqec-backend` to `web` to disambiguate.

> Also, I think that the backend server files should be moved in this PR (per its description). Once this is done, I am happy to validate the changes :)

Merging with `main` re-created some of the server files in `src`, so I'll go ahead and delete those and add that to this PR.

I asked about teams since @nelimee wanted to be included in a previous PR and wanted to suggest a system for making sure people are assigned the PRs that they are interested in participating in.

**Comment by vtomole** [comment]
> I asked about teams since @nelimee wanted to be included in a previous PR 

@nelimee wouldn't have been needed in the previous PR if I had suggested we put that code in the `web` directory. My bad. I don't think teams are needed at the moment since it's only 4 of us. After moving the server files to `web` anything in `src` will need @nelimee as a reviewer.

**Comment by vtomole** [comment]
Now that i think about it, maybe we want to rename  `web` to `server` because `tqecjs` would also qualify as `web`.

**Comment by smburdick** [comment]
> Now that i think about it, maybe we want to rename `web` to `server` because `tqecjs` would also qualify as `web`.

To make it even more descriptive, how about `flask-server`?

I was thinking about `tqecjs/` as well. It'd make sense to keep that name if there were more multiple frontends (for example, if we decided to write a tkinter GUI that could run the whole nine yards locally).

**Comment by vtomole** [comment]
`server` is fine. I don't think we are going to be writing other servers in the future. Likewise for `tqecjs`. We are already short-staffed with respect to creating one frontend. We probably won't be writing others. Maybe we can rename it to `frontend` for now.

**Comment by vtomole** [comment]
@smburdick Thanks. Let's move the discussion renaming of `tqecjs` elsewhere.

**Comment by rryoung98** [comment]
> > Although structurally im a little hesitant to put the post request in the index.js,
> 
> Can you elaborate on where it should be? Would it make more sense to create a `downloadButton` class and invoke it from the `workspace`?

yes I figured given the number of calls we'll make, that it might be good to dedicate a file to api calls, like a "templateActions.js"

**Comment by smburdick** [comment]
> > > Although structurally im a little hesitant to put the post request in the index.js,
> > 
> > 
> > Can you elaborate on where it should be? Would it make more sense to create a `downloadButton` class and invoke it from the `workspace`?
> 
> yes I figured given the number of calls we'll make, that it might be good to dedicate a file to api calls, like a "templateActions.js". not a big deal at the moment :)

@rryoung98 Are there any changes you'd like me to make to the current revision?

**Review by rryoung98 — COMMENTED** [review]
The PR needs to wait a bit because it still works in the plaquette first then tile interface and there have been a decent amount of changes made on my PR https://github.com/QCHackers/tqec/pull/69 that is intended to fix these.

Could we merge my branch then check for merge conflicts on this branch? In particular with the TileClass (now TemplateClass).

Otherwise, changes look good.

**Comment by smburdick** [comment]
> The PR needs to wait a bit because it still works in the plaquette first then tile interface and there have been a decent amount of changes made on my PR #69 that is intended to fix these.
> 
> Could we merge my branch then check for merge conflicts on this branch? In particular with the TileClass (now TemplateClass).
> 
> Otherwise, changes look good.

@rryoung98 ,

Thanks. I'm taking a look at #69 right now. We can merge that one first.

I will admit I went a little out of scope in this PR -- I did a lot of refactoring on Saturday and wanted to push it out. It probably should have been in its own PR, but I was already making some changes to the frontend in this PR anyway, so I added it on anyway.

I think once these are merged, we should follow up on this morning's ask of creating more tasks for the frontend and assigning those, so no work is duplicated. I didn't really lose any time myself -- it was good practice for me anyway :) @vtomole are you interested in helping with the frontend task breakdown too?

**Comment by smburdick** [comment]
> > @vtomole are you interested in helping with the frontend task breakdown too?
> 
> Sure. Is there an existing issue that we can take this discussion to?

I created #74 for planning purposes.

**Comment by rryoung98** [comment]
Hey @smburdick  i think this is safe to commit after fixing the conflicts, let me know if you have any questions on how things should be merged

**Comment by rryoung98** [comment]
Sorry about that, and i can be more diligent about writing issues

**Comment by smburdick** [comment]
Had to close this PR as I have moved my work to another branch.

**Pull request closed without merging**
